File: android/app/src/main/kotlin/com/bockmedia/console/domain/model/TabWarmCoordinator.kt
package com.bockmedia.console.domain.model

import android.content.Context
import com.bockmedia.console.data.repository.BockMediaRepository
import com.bockmedia.console.domain.model.LibraryCachePersistence
import com.bockmedia.console.ui.library.LibraryArtPrefetch
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

/** Deferred tab prefetch — run when a tab is first opened, not on home boot. */
object TabWarmCoordinator {
    fun warmLibrary(scope: CoroutineScope, context: Context, repository: BockMediaRepository) {
        if (LibrarySessionCache.getIfFresh() != null) return
        scope.launch {
            runCatching {
                val data = LibraryLoader.loadBuckets(repository, context)
                LibrarySessionCache.put(data)
                LibraryCachePersistence.save(context, data)
                LibraryArtPrefetch.warm(context, repository, data.forFilter(LibraryFilter.All))
            }
        }
    }

    fun warmSearchBrowse(scope: CoroutineScope, repository: BockMediaRepository) {
        if (SearchBrowseSessionCache.getIfFresh() != null) return
        scope.launch {
            runCatching {
                val fast = SearchBrowseLoader.loadFast(repository)
                SearchBrowseSessionCache.put(fast)
                val full = SearchBrowseLoader.load(repository)
                SearchBrowseSessionCache.put(full)
            }
        }
    }
}

File: ios/BockMedia/Domain/TabWarmCoordinator.swift
import Foundation

/// Deferred tab prefetch — run from Home after first paint, not on cold boot.
enum TabWarmCoordinator {
    static func warmLibrary(repository: BockMediaRepository) {
        if LibrarySessionCache.getIfFresh() != nil { return }
        Task {
            let memberKey = ActiveProfileStore.activeMemberId() ?? ""
            if let cached = SessionDataStore.peekPlaylists(memberKey: memberKey), !cached.items.isEmpty {
                let playlistItems = cached.items.map { pl in
                    LibraryItem(
                        id: "pl-\(pl.id)",
                        title: pl.name,
                        subtitle: "\(pl.tracks) songs",
                        kind: .playlist,
                        playTarget: .playlist(id: pl.id, name: pl.name),
                        artPath: pl.artPath,
                        playlistId: pl.id
                    )
                }
                LibrarySessionCache.put(LibraryData(
                    playlists: playlistItems,
                    artists: [],
                    albums: [],
                    offline: []
                ))
            }
            let data = await LibraryLoader.loadBuckets(repository: repository)
            LibrarySessionCache.put(data)
            LibraryCachePersistence.save(data)
        }
    }

    static func warmSearchBrowse(repository: BockMediaRepository) {
        if SearchBrowseSessionCache.getIfFresh() != nil { return }
        Task {
            let full = await SearchBrowseLoader.load(repository: repository)
            SearchBrowseSessionCache.put(full)
        }
    }
}
